"""
=========================================================
Loss Functions From Scratch (NumPy Implementation)
=========================================================

This module implements core loss (cost) functions used in
Machine Learning and Deep Learning models, written entirely
from scratch using only NumPy.

The objective of a loss function is to measure how far the
model's predictions are from the actual ground-truth values.
Neural networks learn by minimizing this loss during training.

Implemented Loss Functions
--------------------------
1. Mean Squared Error (MSE)
   - Used for regression tasks
   - Penalizes large errors heavily (squared difference)

2. Mean Absolute Error (MAE)
   - Used for regression tasks
   - More robust to outliers than MSE

3. Binary Cross Entropy (BCE)
   - Used for binary classification problems
   - Example: spam detection, disease prediction

4. Categorical Cross Entropy (CCE)
   - Used for multi-class classification
   - Example: digit recognition, image classification

Why this project?
-----------------
High-level libraries like TensorFlow and PyTorch compute the
loss automatically, which hides the underlying mathematics.
This project aims to understand the internal working of neural
networks by manually implementing the loss functions.

Learning Outcome
----------------
By studying this code, one can understand:
• How model error is calculated
• Why different tasks require different loss functions
• The role of loss in gradient descent and backpropagation

This project is part of the "Neural Networks From Scratch"
learning series.

Author: (VIKNESHRAJD)
Language: Python
Dependencies: NumPy only

"""


# Import Library

In [1]:
import numpy as np

# Mean Squared Error (MSE)

**Used in:** Regression problems  
*(Example: House price prediction)*

Mean Squared Error measures the average squared difference between the
true values and the predicted values produced by a model.

It penalizes large errors more heavily because the error is squared.

---

### Formula


$$
\text{MSE} = \frac{1}{n} \sum_{i=1}^{n} \left(y_i^{true} - y_i^{pred}\right)^2
$$


Where:

- \( n \) = number of samples
- \( y_{true} \) = actual (ground truth) value
- \( y_{pred} \) = model predicted value



If predictions are very close to the real values → **MSE is small**  
If predictions are far from the real values → **MSE becomes large**

Because of the square term, big mistakes are punished much more than small mistakes.


In [2]:
def mse(y_true, y_pred):
    y_true = np.array(y_true)
    y_pred = np.array(y_pred)

    loss = np.mean((y_true - y_pred) ** 2)
    return loss


# Example
y_true = [3, -0.5, 2, 7]
y_pred = [2.5, 0.0, 2, 8]

print("MSE:", mse(y_true, y_pred))

MSE: 0.375


The model predictions are close to the true values, therefore the Mean Squared Error is small (0.375), indicating good prediction accuracy

# **Mean Absolute Error (MAE)**


**Used in:** Regression problems

Mean Absolute Error calculates the average absolute difference between
the true values and the predicted values.

Unlike MSE, it does NOT square the error, so large mistakes are not
penalized excessively.



### **Formula**

$$
\text{MAE} = \frac{1}{n} \sum_{i=1}^{n} \left| y_i^{true} - y_i^{pred} \right|
$$



### Where

- $n$ = total number of samples  
- $y_i^{true}$ = actual (ground truth) value  
- $y_i^{pred}$ = predicted value  

>

If prediction error is small → **MAE is small**  
If prediction error is large → **MAE increases linearly**

MAE treats all errors equally, whereas MSE punishes large errors more heavily.


In [3]:
def mae(y_true, y_pred):
    y_true = np.array(y_true)
    y_pred = np.array(y_pred)

    loss = np.mean(np.abs(y_true - y_pred))
    return loss


print("MAE:", mae(y_true, y_pred))


MAE: 0.5


The Mean Absolute Error is 0.5, which means the model's predictions
are off by 0.5 units on average from the true values.


Unlike MSE, MAE does not square the errors, so each mistake contributes
proportionally to the loss. Therefore MAE is less sensitive to outliers

| Loss    | Punishes large error? | Outlier sensitivity |
| ------- | --------------------- | ------------------- |
| **MSE** | Yes (squared)         | High                |
| **MAE** | No (linear)           | Low                 |

MSE cares about big mistakes, MAE cares about average mistakes.


## **Binary Cross Entropy (BCE)**

**Used in:** Binary Classification  
*(Example: spam vs not spam, cancer vs no cancer)*

The model output must be a probability between **0 and 1**.
Therefore, BCE is typically used with a **Sigmoid activation function**.

Binary Cross Entropy measures how far the predicted probability is
from the actual class label (0 or 1).


###  Formula

$$
\text{BCE} = -\frac{1}{n} \sum_{i=1}^{n}
\left[
y_i \log(p_i) + (1 - y_i)\log(1 - p_i)
\right]
$$


**Where**,

- $n$ = number of samples  
- $y_i$ = true label (0 or 1)  
- $p_i$ = predicted probability of class 1

---

### Numerical Stability

If $p = 0$ or $p = 1$, then:

$$
\log(0) = -\infty
$$

This causes the training process to crash (NaN loss).

So in real deep learning implementations we clip the prediction:

$$
p = \text{clip}(p, \varepsilon, 1 - \varepsilon)
$$

where:

$$
\varepsilon = 10^{-15}
$$

This small value prevents the logarithm from becoming undefined.



- Correct prediction with high confidence → **very small loss**
- Wrong prediction with high confidence → **very large loss**

BCE heavily punishes confident wrong predictions.


In [4]:
def binary_cross_entropy(y_true, y_pred):
    y_true = np.array(y_true)
    y_pred = np.array(y_pred)

    epsilon = 1e-15   # prevents log(0)
    y_pred = np.clip(y_pred, epsilon, 1 - epsilon)

    loss = -np.mean(
        y_true * np.log(y_pred) +
        (1 - y_true) * np.log(1 - y_pred)
    )

    return loss


# Example
y_true = [1, 0, 1, 1]
y_pred = [0.9, 0.2, 0.8, 0.7]

print("Binary Cross Entropy:", binary_cross_entropy(y_true, y_pred))

Binary Cross Entropy: 0.22708064055624455


The Binary Cross Entropy loss is 0.227.


This indicates that the predicted probabilities are close to the
true labels. A lower BCE value means the classifier is making
accurate predictions with high confidence.


If the model predicted the wrong class with high confidence,
the loss would become very large.



For Binary Cross Entropy, the loss depends on the true label.


**If the true label is 1:**

$$
\text{Loss} = -\log(p)
$$

(The model should predict a probability close to 1)

---

**If the true label is 0:**

$$
\text{Loss} = -\log(1 - p)
$$

(The model should predict a probability close to 0)



- If the prediction is correct with high confidence → loss is very small
- If the prediction is wrong with high confidence → loss becomes very large

This is why Binary Cross Entropy is ideal for classification problems.


| Prediction      | True Label | Loss       |
| --------------- | ---------- | ---------- |
| p = 0.9 , y = 1 | correct    | very small |
| p = 0.1 , y = 1 | wrong      | very large |


## **Categorical Cross Entropy (CCE)**

**Used in:** Multi-class classification  
*(Example: digit recognition 0–9, cat/dog/horse classifier)*

In multi-class problems, the model predicts a probability for each class.
Therefore:

- $y_{true}$ must be **one-hot encoded**
- $y_{pred}$ must be **Softmax probabilities**

Categorical Cross Entropy measures how close the predicted probability
distribution is to the true distribution.


### Formula

$$
\text{CCE} = - \sum_{i=1}^{C} y_i^{true} \log(y_i^{pred})
$$

For a dataset with multiple samples:

$$
\text{CCE} = -\frac{1}{n} \sum_{j=1}^{n} \sum_{i=1}^{C}
y_{ij}^{true} \log(y_{ij}^{pred})
$$


> **Where**

- $n$ = number of samples  
- $C$ = number of classes  
- $y_{ij}^{true}$ = true label (one-hot encoded)  
- $y_{ij}^{pred}$ = predicted probability from Softmax  


> **Example**: One-Hot Encoding

If there are 4 classes and the correct class is **class 2**:

$$
[0,\; 0,\; 1,\; 0]
$$

The model might predict:

$$
[0.1,\; 0.2,\; 0.6,\; 0.1]
$$

Since the true class index is the 3rd position, the loss becomes:

$$
\text{Loss} = -\log(0.6)
$$


- If the model gives **high probability** to the correct class → small loss
- If the model gives **low probability** to the correct class → large loss

CCE encourages the network to assign maximum probability to the correct class.


In [8]:
def categorical_cross_entropy(y_true, y_pred):
    y_true = np.array(y_true)
    y_pred = np.array(y_pred)

    epsilon = 1e-15
    y_pred = np.clip(y_pred, epsilon, 1 - epsilon)

    loss = -np.mean(np.sum(y_true * np.log(y_pred), axis=1))
    return loss


# Example (3 classes)
y_true = [
    [1, 0, 0],
    [0, 1, 0],
    [0, 0, 1]
]

y_pred = [
    [0.8, 0.1, 0.1],
    [0.2, 0.7, 0.1],
    [0.1, 0.2, 0.7]
]

print("Categorical Cross Entropy:", categorical_cross_entropy(y_true, y_pred))


Categorical Cross Entropy: 0.3121644797305582


The Categorical Cross Entropy loss is 0.312.

The model is assigning high probability to the correct class for
each sample, therefore the loss is low. If the model predicted
a low probability for the correct class, the loss would increase
significantly.


In multi-class classification, Categorical Cross Entropy only cares about the probability assigned to the correct class, not the incorrect ones.

### Understanding the CCE Output

In **Categorical Cross Entropy**, only the probability assigned to the
**correct class** affects the loss.

For each sample:

$$
\text{Loss} = -\log(p_{correct})
$$



> ### Per-sample calculation

**1st sample** → correct class probability = 0.8

$$
-\log(0.8) = 0.2231
$$

**2nd sample** → correct class probability = 0.7

$$
-\log(0.7) = 0.3567
$$

**3rd sample** → correct class probability = 0.7

$$
-\log(0.7) = 0.3567
$$


> ### Average loss

$$
\text{CCE} =
\frac{0.2231 + 0.3567 + 0.3567}{3}
= 0.3122
$$

This matches the computed output:

Categorical Cross Entropy ≈ 0.31216

| Loss                      | Use Case                        | Output Activation |
| ------------------------- | ------------------------------- | ----------------- |
| MSE                       | Regression                      | Linear            |
| MAE                       | Regression (robust to outliers) | Linear            |
| Binary Cross Entropy      | 2 classes                       | Sigmoid           |
| Categorical Cross Entropy | Multi-class                     | Softmax           |
